gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/filter/BinaryReductionStep.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.tinkerpop.gremlin.process.traversal.step.filter;
+
+import org.apache.tinkerpop.gremlin.process.traversal.GremlinTypeErrorException;
+
+/**
+ * Any {@link FilterStep} marked as a BinaryReductionStep will catch {@link GremlinTypeErrorException}s produced by
```

Comment:
I am a bit apprehensive about maintainability of BinaryReductionStep. If a new step is added in future, it would be upto the implementors good intentions to add this interface. Instead, can we make it more explicit? 

Some possible ideas:
1. Do we need this interface? Can we say that all FilterSteps convert `GremlinTypeErrorException` into a boolean?What are the cases where they don't?
2. Why do we wait to convert this exception into false at the top level? Why not convert it in the lowest granularity step itself?

Reply:
1. Child traversals need an opportunity to reduce from ternary to binary. We can't say all FilterSteps convert, because of (2) propagation.
2. We need to propagate `ERROR` up to the last minute before reducing to `FALSE`, since it can be combined in interesting ways via `AND/OR/NOT`.

It seems extremely unlikely to me there would be another step added that would need to worry about this. Any normal filter step would not. It's a quasi-top level filter like .where and .filter that might need to worry about binary reduction in the context of a child traversal.

Reply:
Understood. Resolving this thread.